protobuf_serialization/internal.nim

```diff
@@ -0,0 +1,156 @@
+  CastableLengthDelimitedTypes* = string or seq[char or byte or uint8 or bool]
+  #This type is literally every other type.
+  #Every other type is considered custom, due to the need for their own converters.
+  #While cstring/array are built-ins, and therefore should have converters provided, but they still need converters.
+  LengthDelimitedTypes* = not (VarIntTypes or FixedTypes)
+
+proc getTypeChain(impure: NimNode): seq[NimNode] {.compileTime.} =
+  var current = impure
+  result = @[]
+  while true:
+    let
+      impl = getTypeImpl(current)
+      inst = getTypeInst(current)
+    if (impl.kind == nnkBracketExpr) and (impl[0].strVal == "typeDesc"):
+      current = impl[1]
+    elif impl.kind == nnkRefTy:
+      result.add(newNimNode(nnkRefTy))
+      result[^1].add(inst)
+      current = impl[0]
+    elif inst.kind == nnkSym:
+      if (result.len == 0) or (
+        not (
+          (result[^1].kind == nnkSym) and
+          (result[^1].strVal == inst.strVal)
+        )
+      ):
+        result.add(inst)
+      break
+    elif (inst.kind == nnkBracketExpr) and (inst[0].kind == nnkSym) and (inst[0].strVal == "Option"):
+      current = inst[1]
+      result.add(newNimNode(nnkBracketExpr))
+    elif inst.kind == nnkBracketExpr:
+      result.add(inst)
+      break
+    else:
+      break
+
+macro isPotentiallyNull*(impure: typed): bool =
+  for ty in getTypeChain(impure):
+    if (ty.kind == nnkRefTy) or (ty.kind == nnkBracketExpr):
+      return newLit(true)
+  result = newLit(false)
+
+macro flatType*(impure: typed): untyped =
+  getTypeChain(impure)[^1]
+
+proc flatMapInternal[T, B](value: T, base: typedesc[B]): Option[B] =
+  when value is Option:
+    if value.isNone():
+      none(base)
+    else:
+      flatMapInternal(value.get(), base)
+  elif value is ref:
+    if value.isNil:
+      none(base)
+    else:
+      flatMapInternal(value[], base)
+  else:
+    some(value)
+
+macro flatMap*(value: typed): untyped =
+  let flattened = getTypeChain(value)[^1]
+  quote do:
+    flatMapInternal(`value`, type(`flattened`))
+
+macro box*(variable: typed, value: typed): untyped =
+  let chain = getTypeChain(variable)
+  var wrap = value
+  for l in countdown(high(chain) - 1, 0):
+    if chain[l].kind == nnkRefTy:
+      wrap = newBlockStmt(
+        newStmtList(
+          newNimNode(nnkVarSection).add(
+            newNimNode(nnkIdentDefs).add(
+              ident("newRef"),
+              newNimNode(nnkEmpty),
+              newNimNode(nnkCall).add(chain[l][0])
+            )
+          ),
+          newNimNode(nnkAsgn).add(
+            newNimNode(nnkBracketExpr).add(
+              ident("newRef")
+            ),
+            wrap
+          ),
+          ident("newRef")
+        )
+      )
+    elif chain[l].kind == nnkBracketExpr:
+      wrap = newNimNode(nnkCall).add(ident("some"), wrap)
+
+  quote do:
+    `variable` = `wrap`
+
+macro isStdlib*(ty: untyped): untyped =
```

Comment:
This is definetely not needed. What's wrong with using the `is` operator directly everywhere where `isStdLib` is used and having simpler checks such as:

```nim
when value is seq:
   writeSeq(value)
elif value is set:
   writeSet(value)
...
```